_posts/2019-08-01-my-own-summer-neural-network.md

```diff
@@ -0,0 +1,321 @@
+```
+
+From this data, I get predictions in a form of `[go_left_prob, go_right_prob, go_forward_prob]`.
+
+### 4. Enter the TensorFlow
+
+This is the part of my experiment that I have absolutely no knowledge of. I just simply copied the neural network used in GTA examples and I learned only one thing:
+the input and output have shapes which must be defined in the network.
+
+This is the neural network configuration used, but I have absolutely no idea, what it actually does :D
+
+```python
+# alexnet.py
+
+# Shamesly copied from Python plays GTA videoseries
+# https://www.youtube.com/playlist?list=PLQVvvaa0QuDeETZEOy4VdocT7TOjfSA8a
+
+""" AlexNet.
+References:
+    - Alex Krizhevsky, Ilya Sutskever & Geoffrey E. Hinton. ImageNet
+    Classification with Deep Convolutional Neural Networks. NIPS, 2012.
+Links:
+    - [AlexNet Paper](http://papers.nips.cc/paper/4824-imagenet-classification-with-deep-convolutional-neural-networks.pdf)
+"""
+
+import tflearn
+from tflearn.layers.conv import conv_2d, max_pool_2d
+from tflearn.layers.core import input_data, dropout, fully_connected
+from tflearn.layers.estimator import regression
+from tflearn.layers.normalization import local_response_normalization
+
+def alexnet(width, height, lr):
+    network = input_data(shape=[None, width, height, 1], name='input')
+    network = conv_2d(network, 96, 11, strides=4, activation='relu')
+    network = max_pool_2d(network, 3, strides=2)
+    network = local_response_normalization(network)
+    network = conv_2d(network, 256, 5, activation='relu')
+    network = max_pool_2d(network, 3, strides=2)
+    network = local_response_normalization(network)
+    network = conv_2d(network, 384, 3, activation='relu')
+    network = conv_2d(network, 384, 3, activation='relu')
+    network = conv_2d(network, 256, 3, activation='relu')
+    network = max_pool_2d(network, 3, strides=2)
+    network = local_response_normalization(network)
+    network = fully_connected(network, 4096, activation='tanh')
+    network = dropout(network, 0.5)
+    network = fully_connected(network, 4096, activation='tanh')
+    network = dropout(network, 0.5)
+    network = fully_connected(network, 3, activation='softmax')
+    network = regression(network, optimizer='momentum',
+                         loss='categorical_crossentropy',
+                         learning_rate=lr, name='targets')
+
+    model = tflearn.DNN(network, checkpoint_path='model_alexnet',
+                        max_checkpoints=1, tensorboard_verbose=2, tensorboard_dir='log')
+
+    return model
+```
+
+Then I just ran some code that fits the data to the neural network and to my surprise, it worked and TensorFlow produced the model files!
+
+### 5. Implement the Agent
+
+Implementing the agent is pretty similar to implementing the capturing mechanism, but instead of capturing the current keyboard state, you ask prediction for the image from neural network and use the given result to send a keyboard event.
```

Comment:
```suggestion
Implementing the agent is pretty similar to implementing the capturing mechanism, but instead of capturing the current keyboard state, you ask for a prediction for the image from the neural network and use the given result to send a keyboard event.
```